packages/shadydom/src/patch-events.js

```diff
@@ -306,48 +306,50 @@ function fireHandlers(event, node, phase) {
 }
 
 function shadyDispatchEvent(e) {
-  let path = e.composedPath();
-  let node;
+  const path = e.composedPath();
+  const retargetedPath = path.map(node => retarget(node, path));
+  const bubbles = e.bubbles;
+
+  let currentTarget;
   // override `currentTarget` to let patched `target` calculate correctly
   Object.defineProperty(e, 'currentTarget', {
     get: function() {
-      return node;
+      return currentTarget;
     },
     configurable: true
   });
+
+  let eventPhase = Event.CAPTURING_PHASE;
+  Object.defineProperty(e, 'eventPhase', {
+    get: function() {
+      return eventPhase;
+    },
+  });
+
   for (let i = path.length - 1; i >= 0; i--) {
-    node = path[i];
+    currentTarget = path[i];
+    eventPhase = currentTarget === retargetedPath[i] ? Event.AT_TARGET : Event.CAPTURING_PHASE;
     // capture phase fires all capture handlers
-    fireHandlers(e, node, 'capture');
+    fireHandlers(e, currentTarget, 'capture');
     if (e.__propagationStopped) {
       return;
     }
   }
 
-  // set the event phase to `AT_TARGET` as in spec
-  Object.defineProperty(e, 'eventPhase', {get() { return Event.AT_TARGET }});
-
-  const bubbles = e.bubbles;
-  // the event only needs to be fired when owner roots change when iterating the event path
-  // keep track of the last seen owner root
-  let lastFiredRoot;
   for (let i = 0; i < path.length; i++) {
-    node = path[i];
-    const nodeData = shadyDataForNode(node);
-    const root = nodeData && nodeData.root;
-    if (!(i === 0 || (root && root === lastFiredRoot)) && !bubbles) {
-      continue;
-    }
-
-    fireHandlers(e, node, 'bubble');
-    // don't bother with window, it doesn't have `getRootNode` and will be last in the path anyway
-    if (node !== window) {
-      lastFiredRoot = node[utils.SHADY_PREFIX + 'getRootNode']();
-    }
-    if (e.__propagationStopped) {
-      return;
+    currentTarget = path[i];
+    const atTarget = currentTarget === retargetedPath[i];
+    if (atTarget || bubbles) {
```

Comment:
Can this loop be broken out of when `atTarget` and `bubbles` are false?

Reply:
Not always: if the event is non-bubbling, composed, and dispatched to an element in a shadow root, then both the real target and the shadow root host will see `atTarget` as true even though there are other items in `path` between them.